core/src/main/java/org/infinispan/interceptors/impl/PrefetchInterceptor.java

```diff
@@ -207,68 +214,97 @@ private Object prefetchKeysIfNeededAndInvokeNext(InvocationContext ctx, Visitabl
-         for (Response response : responseMap.values()) {
-            if (!response.isSuccessful()) {
-               throw OutdatedTopologyException.INSTANCE;
-            }
-            SuccessfulResponse successfulResponse = (SuccessfulResponse) response;
-            InternalCacheValue icv = (InternalCacheValue) successfulResponse.getResponseValue();
-            if (icv == null) {
-               continue;
-            }
-            Metadata metadata = icv.getMetadata();
-            if (metadata instanceof RemoteMetadata) {
-               // not sure if this can happen, but let's be on the safe side
-               throw OutdatedTopologyException.INSTANCE;
-            }
-            if (metadata != null && metadata.version() != null) {
-               if (maxVersion == null || maxVersion.compareTo(metadata.version()) == InequalVersionComparisonResult.BEFORE) {
-                  maxVersion = metadata.version();
-                  maxValue = icv;
-               }
+      command.setTopologyId(topologyId);
+      return rpcManager.invokeRemotelyAsync(targets, command, syncRpcOptions).thenApply(handleRemotelyPrefetchedEntry);
+   }
+
+   private InternalCacheValue handleRemotelyPrefetchedEntry(Map<Address, Response> responseMap) {
+      EntryVersion maxVersion = null;
+      InternalCacheValue maxValue = null;
+      for (Response response : responseMap.values()) {
+         if (!response.isSuccessful()) {
+            throw OutdatedTopologyException.INSTANCE;
+         }
+         SuccessfulResponse successfulResponse = (SuccessfulResponse) response;
+         InternalCacheValue icv = (InternalCacheValue) successfulResponse.getResponseValue();
+         if (icv == null) {
+            continue;
+         }
+         Metadata metadata = icv.getMetadata();
+         if (metadata instanceof RemoteMetadata) {
+            // not sure if this can happen, but let's be on the safe side
+            throw OutdatedTopologyException.INSTANCE;
+         }
+         if (metadata != null && metadata.version() != null) {
+            if (maxVersion == null || maxVersion.compareTo(metadata.version()) == InequalVersionComparisonResult.BEFORE) {
+               maxVersion = metadata.version();
+               maxValue = icv;
             }
          }
-         return maxValue;
-      });
+      }
+      if (trace) {
+         log.tracef("Prefetched value is %s", maxValue);
+      }
+      return maxValue;
    }
 
-   private InvocationStage retrieveRemoteValues(InvocationContext ctx, List<?> keys) {
+   // TODO: this is not completely aligned with single-entry prefetch
+   private InvocationStage retrieveRemoteValues(InvocationContext ctx, List<?> keys, int topologyId) {
+      if (trace) {
+         log.tracef("Prefetching entries for keys %s using broadcast", keys);
+      }
       ClusteredGetAllCommand command = commandsFactory.buildClusteredGetAllCommand(keys, FlagBitSets.SKIP_OWNERSHIP_CHECK, null);
+      command.setTopologyId(topologyId);
```

Comment:
It's just odd to have such a long explanation in one method and no explanation in the another one when they do the same thing :)

TBH it's probably better to not have a comment in `PrefetchInterceptor`, and have `ScatteredDistributionInterceptor` throw an exception if the topology id is not set. In https://github.com/infinispan/infinispan/pull/5315 I also made `RpcManager.invokeCommand*()` throw an exception if the topology id is not set instead of using the current topology id, like `RpcManager.invokeRemotely()` does.